### PR TITLE
Make RabbitTemplate exchange and routingKey configurable

### PR DESCRIPTION
**Pull request by ArloL**
This adds three properties to `RabbitProperties.Template` and sets them on the auto-configured `RabbitTemplate` bean. These are the three properties our internal code sets the most and would love to be able to control them via external configuration.

Please let me know if there are any issues.

Thanks for the great work on Spring Boot!

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java [hunk 1]
```diff
@@ -699,6 +699,21 @@ public void setConsumersPerQueue(Integer consumersPerQueue) {
 		 */
 		private Long replyTimeout;
 
+		/**
+		 * The name of the default exchange to use for send operations.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java]: Configuration properties description do not start with"The", "A", etc. Please look at the appendix to harmonize them.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java [hunk 2]
```diff
@@ -699,6 +699,21 @@ public void setConsumersPerQueue(Integer consumersPerQueue) {
...
+		 */
+		private String exchange;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java]: The default exchange is `""` so it would be nice if the property would say that rather than `null`

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java [hunk 3]
```diff
@@ -699,6 +699,21 @@ public void setConsumersPerQueue(Integer consumersPerQueue) {
...
+
+		/**
+		 * The value of a default routing key to use for send operations.
+		 */
+		private String routingKey;
+
+		/**
+		 * Enable transactional channels.
+		 */
+		private boolean channelTransacted;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java]: This one looks totally unrelated to the the first two and makes me a bit neverous. 

@garyrussell, any advice here? Does that sound safe to change the default via properties?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java [hunk 4]
```diff
@@ -186,6 +186,13 @@ public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
 			if (templateProperties.getReplyTimeout() != null) {
 				rabbitTemplate.setReplyTimeout(templateProperties.getReplyTimeout());
 			}
+			if (templateProperties.getExchange() != null) {
+				rabbitTemplate.setExchange(templateProperties.getExchange());
+			}
+			if (templateProperties.getRoutingKey() != null) {
+				rabbitTemplate.setRoutingKey(templateProperties.getRoutingKey());
+			}
+			rabbitTemplate.setChannelTransacted(templateProperties.isChannelTransacted());
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java]: A test should cover this. If we go with the route of flagging `""` a the default value. We should have a test that this value is consistent with what happens  by default.

**Comment by ArloL** [comment]
Thanks for the feedback. I have reviewed your suggestions.

The default values now match those of RabbitTemplate. A test ensures this. I am not sure whether the test is the way you would want it though. Let me know if you want me to refactor it.

I have removed "The" from the descriptions to match the existing style.

If you have any further suggestions let me know and I'll be happy to review them.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java [hunk 5]
```diff
@@ -188,6 +188,9 @@ public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
 			if (templateProperties.getReplyTimeout() != null) {
 				rabbitTemplate.setReplyTimeout(templateProperties.getReplyTimeout());
 			}
+			rabbitTemplate.setExchange(templateProperties.getExchange());
+			rabbitTemplate.setRoutingKey(templateProperties.getRoutingKey());
+			rabbitTemplate.setChannelTransacted(templateProperties.isChannelTransacted());
```
Comment (garyrussell) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java]: I would prefer the use of null checks rather than "mirroring" the current defaults in the template - if we ever change the defaults in the template they will be undone here. See the properties above (reply timeout etc) for examples; so `channelTransacted` should be `Boolean`  so we can check whether the user has explicitly set the property.

While it is unlikely those defaults will change, this is the general pattern used by Boot's auto configuration.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java (hunk 3 above)
Comment (garyrussell) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java]: Routing key and exchange should default to `null` and make this a `Boolean` - see my other comment.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java (hunk 5 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java]: > this is the general pattern used by Boot's auto configuration.

Actually, that's not 100% accurate. When we can add a test  that validates that our default are the same as the underlying tech's default, we try as much as we can to put it back in the config. That way, users can know what the defaults are in the IDE.

Comment (garyrussell) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java]: OK; that makes sense.

As discussed offline; if we expose the `channelTransacted` flag on the template, we should do the same for the listener container factories so that any rabbit template operations performed on the container thread will participate in the same transaction.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java (hunk 3 above)
Comment (ArloL) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitProperties.java]: I will leave this as `boolean` and `false` then since the value is tested.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java (hunk 5 above)
Comment (ArloL) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java]: I will leave these then since they are tested.

**Comment by ArloL** [comment]
I have rebased the PR to trigger another CI build since the last one failed due to an unrelated module.

**Comment by snicoll** [comment]
@ArloL thanks for the PR. I've polished your commit in 6a4a0e3 (in particular adding a test for the feature that was added). As discussed in the review, I have not added the transacted flag as it makes me nervous to add it without some sort of coordination with the message listener.

I think we need to discuss the various cases before adding this feature. Feel free to open a separate issue once you've a more precise idea.

**Comment by habuma** [comment]
@snicoll : Just noticed this PR (and like it a lot). Was wondering if there's any reason why there's not a similar means of setting the default queue for the purposes of `RabbitTemplate.receive()` methods that rely on a default queue? Or am I overlooking something?

I'll be happy to submit a PR when I get a chance (I'm on the road ATM or else I'd just do it now) if this is something that's both missing and makes sense to you.
